pkg/ingester/client/client.go

```diff
@@ -94,21 +136,110 @@ func MakeIngesterClient(addr string, cfg Config) (HealthAndIngesterClient, error
 	if err != nil {
 		return nil, err
 	}
-	return &closableHealthAndIngesterClient{
+	c := &closableHealthAndIngesterClient{
 		IngesterClient:          NewIngesterClient(conn),
 		HealthClient:            grpc_health_v1.NewHealthClient(conn),
 		conn:                    conn,
 		addr:                    addr,
 		maxInflightPushRequests: cfg.MaxInflightPushRequests,
 		inflightPushRequests:    ingesterClientInflightPushRequests,
-	}, nil
+	}
+	if useStreamConnection {
+		streamCtx, streamCancel := context.WithCancel(context.Background())
+		err = c.Run(make(chan *streamWriteJob, INGESTER_CLIENT_STREAM_WORKER_COUNT), streamCtx, streamCancel)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
 }
 
 func (c *closableHealthAndIngesterClient) Close() error {
 	c.inflightPushRequests.DeleteLabelValues(c.addr)
+
+	if c.streamCancel != nil {
+		c.streamCancel()
+	}
+
+	if c.streamPushChan != nil {
+	drainingLoop:
+		for {
+			select {
+			case job, ok := <-c.streamPushChan:
+				if !ok {
+					break drainingLoop
+				}
+				if job != nil && job.cancel != nil {
+					job.err = errors.New("stream connection ingester client closing")
+					job.cancel()
+				}
+			default:
+				close(c.streamPushChan)
+				break drainingLoop
+			}
+		}
+	}
+
 	return c.conn.Close()
 }
 
+func (c *closableHealthAndIngesterClient) Run(streamPushChan chan *streamWriteJob, streamCtx context.Context, streamCancel context.CancelFunc) error {
+	c.streamPushChan = streamPushChan
+	c.streamCtx = streamCtx
+	c.streamCancel = streamCancel
+	var err error
+	for i := 0; i < INGESTER_CLIENT_STREAM_WORKER_COUNT; i++ {
+		go func() {
+			for {
+				select {
+				case <-streamCtx.Done():
+					return
+				default:
+					err = c.worker(streamCtx)
```

Comment:
Does this have race condition when setting error? err variable is shared and multiple goroutines can modify it at the same time. Maybe better to use the pattern of error channel